mono/metadata/threads.c

```diff
@@ -464,35 +502,123 @@ static void ensure_synch_cs_set (MonoInternalThread *thread)
 	synch_cs = g_new0 (MonoCoopMutex, 1);
 	mono_coop_mutex_init_recursive (synch_cs);
 
-	if (mono_atomic_cas_ptr ((gpointer *)&thread->synch_cs,
-					       synch_cs, NULL) != NULL) {
+	gpointer ret = mono_atomic_cas_ptr ((gpointer *)&thread->synch_cs, synch_cs, NULL);
+
+	if (ret != NULL) {
 		/* Another thread must have installed this CS */
+		/* Or the finalizer had to have run */
 		mono_coop_mutex_destroy (synch_cs);
 		g_free (synch_cs);
 	}
 }
 
-static inline void
-lock_thread (MonoInternalThread *thread)
+static void
+internal_thread_free (MonoInternalThread *thread)
 {
+	MonoCoopMutex *synch_cs = thread->synch_cs;
+	thread->synch_cs = INTERNAL_THREAD_TOMBSTONE;
+	mono_memory_barrier ();
+
+	mono_coop_mutex_destroy (synch_cs);
+	g_free (synch_cs);
+}
+
+static void
+cleanup_thread_lock (MonoInternalThread *thread)
+{
+	thread->freeing = 1;
+	mono_memory_barrier ();
+
+	gint32 waiters = mono_atomic_load_i32 (&thread->num_waiters);
+
+	// If we started cleanup without any writers, finish the freeing
+	// ourselves. Else the thread who unlocks the thread after the last pending 
+	// wait is finished will do the freeing.
+	if (waiters == 0)
+		internal_thread_free (thread);
+}
+
+static void
+decrement_wait_queue (MonoInternalThread *thread)
+{
+	gint32 waiters = mono_atomic_dec_i32 (&thread->num_waiters);
+	if (waiters == 0) {
+		gboolean freeing = mono_atomic_load_i32 (&thread->freeing);
+		if (freeing)
```

Comment:
What guarantees that there are no new waiters at this point?

Reply:
The ordering of events ensures that new waiters will not be added after the thread enters "freeing" mode.